### PR TITLE
I2C_exercise_code

### PR DESCRIPTION
**Pull request by Mirabellensaft**
This pr contains the exercise code solution. It still needs to be turned into gap texts.

### CODE REVIEW

File: advanced/i2c-driver-exercise/src/imc42670p.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    AD0 = 0b110_1000,
+    AD1 = 0b110_1001,
+}
+
+impl<I2C, E>IMC42670P<I2C>
+where
+    I2C: i2c::WriteRead<Error = E> + i2c::Write<Error = E>,
+{
+    /// Add doc comments!
+    pub fn new(i2c: I2C, address: SlaveAddr) -> Result<Self, E> {
```
Comment (jonathanpallant) [inline advanced/i2c-driver-exercise/src/imc42670p.rs:29]: We should explain here that if you own the I2C device, no-one else can use the bus whilst this driver exists. The users options are then either
* live with this, and create and destroy the driver every time they want to talk to the sensor,
* take the I2C bus as an argument to the `read_XXX` functions,
* use https://docs.rs/shared-bus/latest/shared_bus/.

File: advanced/i2c-driver-exercise/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    let i2c = Master::<I2C0, _, _>::new(
+        peripherals.i2c0,
+        MasterPins { sda, scl },
+        <MasterConfig as Default>::default().baudrate(400.kHz().into()),
+    )?;
+
+    let mut sensor = IMC42670P::new(i2c, SlaveAddr::AD1)?;
+    println!("Sensor init");
+    let serial_num = sensor.read_device_id_register()?;
```
Comment (jonathanpallant) [inline advanced/i2c-driver-exercise/src/main.rs:28]: s/serial_num/device_id/.

**Review by jonathanpallant — CHANGES_REQUESTED** [review]
Generally this looks fine, modulo one suggested replacement of a variable name.

I was a bit confused though as to which bits are skeletons you are expecting people to complete, and what the complete driver over in 'common' was for. Some README files (or file-level comments) would really help!

File: advanced/i2c-driver-exercise/src/imc42670p.rs (hunk 1 above)
Comment (Mirabellensaft) [reply advanced/i2c-driver-exercise/src/imc42670p.rs:29]: I have not made the skeletons yet, one reason is a priority issue and the second is that I wanted someone to look over the solutions I expect to not have to change the same code in multiple places. I added a readme to the advanced folder that explains the purpose of the files.

**Pull request merged**